<a href="https://colab.research.google.com/github/dondreojordan/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module2-make-features/Jordan_Dondre_TL_Group_Serina_DS17.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200> 

# Assignment:

- Replicate the lesson code.

 - This means that if you haven't followed along already, type out the things that we did in class. Forcing your fingers to hit each key will help you internalize the syntax of what we're doing. Make sure you understand each line of code that you're writing, google things that you don't fully understand.
 - [Lambda Learning Method for DS - By Ryan Herr](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit?usp=sharing)
- Convert the `term` column from string to integer.
- Make a column named `loan_status_is_great`. It should contain the integer 1 if `loan_status` is "Current" or "Fully Paid." Else it should contain the integer 0.
- Make `last_pymnt_d_month` and `last_pymnt_d_year` columns.

In [0]:
##### Begin Working Here #####

In [0]:
#URL code to replicate
'https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip'

'https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip'

In [0]:
#Feature Engineering, the process of using a combination of domain knowledge, 
#creativity and the pre-existing columns of a dataset to create new columns.

#Columns of a dataframe each hold specific type of data. 
#We are going to dive in and practice with some common datatypes found in datasets 
#and then we'll make a new feature on a datset using pre-existing conditions. 
#Lets get started. 

import pandas as pd

#Pandas Display Options
#Python will do us a favor and not display all of the categories
#When we are reviewing a dataset. 

#The following code will allow you to set parameters to view the Ames Iowa Dataset:
pd.set_option('display.max_rows', 150)
pd.set_option('display.max_columns', 100)

Parameters are now set to view the data set with the following conditions. Rows=150 max, Columns=100 max

In [0]:
#Lending Club Dataset.
url ='https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip'


#Use the read_csv() method to read dataset.
df = pd.read_csv('https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip')

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0,1,2,3,4,7,13,18,19,24,25,27,28,29,30,31,32,34,36,37,38,39,40,41,42,43,44,46,49,50,51,53,54,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,113,114,115,116,117,118,119,120,121,126,127,131,132,134,135,136,141,142,143) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [0]:

!curl 'https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip'

#Warning message appeared because file could be faily large to display. 

In [0]:
#Dataset is contained in a .zip. 
#Here is a new dataset from the LendingClub 
#that holds information about loans issued in Quarter 4 of 2018.

#the !wget shell command does the same thing that your browser does.
#The only difference is it isn't a webpage, the output is a compressed CSV file.

!wget 'https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip'

--2020-06-02 22:23:10--  https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip
Resolving resources.lendingclub.com (resources.lendingclub.com)... 44.228.179.66, 35.163.140.110, 52.89.160.109
Connecting to resources.lendingclub.com (resources.lendingclub.com)|44.228.179.66|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘LoanStats_2018Q4.csv.zip’

LoanStats_2018Q4.cs     [          <=>       ]  22.10M  1.70MB/s    in 12s     

2020-06-02 22:23:24 (1.83 MB/s) - ‘LoanStats_2018Q4.csv.zip’ saved [23170772]



Now that we are seeing the uncompressed file. 
We need to use the !unzip command to extract the csv from the zipped folder.



> Below you will find steps to unzip and extract data.
>Look at the !head



In [0]:
!unzip LoanStats_2018Q4.csv.zip

Archive:  LoanStats_2018Q4.csv.zip
  inflating: LoanStats_2018Q4.csv    


In [0]:
#In the lecture, we can also use bash/shell 
#commands to look at the raw file using the !head and !tail.

In [0]:
#To look the head and tail of the.csv file:
#  copy and paste the .csv file from 'inflating: " above.
# Do not confuse yourself with the Archive .zipn file


!head LoanStats_2018Q4.csv
# !head method 

Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
"id","member_id","loan_amnt","funded_amnt","funded_amnt_inv","term","int_rate","installment","grade","sub_grade","emp_title","emp_length","home_ownership","annual_inc","verification_status","issue_d","loan_status","pymnt_plan","url","desc","purpose","title","zip_code","addr_state","dti","delinq_2yrs","earliest_cr_line","inq_last_6mths","mths_since_last_delinq","mths_since_last_record","open_acc","pub_rec","revol_bal","revol_util","total_acc","initial_list_status","out_prncp","out_prncp_inv","total_pymnt","total_pymnt_inv","total_rec_prncp","total_rec_int","total_rec_late_fee","recoveries","collection_recovery_fee","last_pymnt_d","last_pymnt_amnt","next_pymnt_d","last_credit_pull_d","collections_12_mths_ex_med","mths_since_last_major_derog","policy_code","application_type","annual_inc_joint","dti_joint","verification_status_joint","acc_now_delinq","tot_coll_amt","tot_cur_bal","open_acc_6m","open_act_il","op

In [0]:
!tail LoanStats_2018Q4.csv
#!tail meathod.

"","","5600","5600","5600"," 36 months"," 13.56%","190.21","C","C1","","n/a","RENT","15600","Not Verified","Oct-2018","Current","n","","","credit_card","Credit card refinancing","836xx","ID","15.31","0","Aug-2012","0","","97","9","1","5996","34.5%","11","w","2926.62","2926.62","3603.44","3603.44","2673.38","930.06","0.0","0.0","0.0","May-2020","190.21","Jun-2020","May-2020","0","","1","Individual","","","","0","0","5996","0","0","0","1","20","0","","0","2","3017","35","17400","1","0","0","3","750","4689","45.5","0","0","20","73","13","13","0","13","","20","","0","3","5","4","4","1","9","10","5","9","0","0","0","0","100","25","1","0","17400","5996","8600","0","","","","","","","","","","","","N","","","","","","","","","","","","","","","N","","","","","",""
"","","23000","23000","23000"," 36 months"," 15.02%","797.53","C","C3","Tax Consultant","10+ years","MORTGAGE","75000","Source Verified","Oct-2018","Charged Off","n","","","debt_consolidation","Debt consolidation","352xx","AL","20.9

As we look at the raw file itself, is there anything that you see that might cause trouble when we read in the CSV file to a dataframe???
 #The structure of the dataset really makes a difference especially when there is added supporting information located within the file. 

In [0]:
#Reading in the CSV.


df = pd.read_csv('LoanStats_2018Q4.csv', header=1)
# Remember: Notice how we still used the .csv and NOT the .zip. 

# After extracting the data now we can use the .head method
#Panda.read_csv() has now taken the data and it is tabular.

df. head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  NaN        NaN    20000.0      20000.0          20000.0   36 months   
1  NaN        NaN    16000.0      16000.0          16000.0   36 months   
2  NaN        NaN     8425.0       8425.0           8425.0   36 months   
3  NaN        NaN    10000.0      10000.0          10000.0   60 months   
4  NaN        NaN    30000.0      30000.0          30000.0   60 months   

  int_rate  installment grade sub_grade                        emp_title  \
0   14.47%       688.13     C        C2                       bus driver   
1   14.47%       550.51     C        C2                        Carpenter   
2   27.27%       345.18     E        E5  Senior Director Risk Management   
3   12.98%       227.43     B        B5                          Printer   
4   26.31%       903.73     E        E4                      Casino Host   

  emp_length home_ownership  annual_inc verification_status   issue_d  \
0    4 years            OWN     52000.0     Source Verified  Dec-2018   
1  10+ years           RENT    150000.0     Source Verified  Dec-2018   
2    3 years       MORTGAGE    450000.0            Verified  Dec-2018   
3    9 years           RENT     60000.0        Not Verified  Dec-2018   
4  10+ years            OWN     62000.0     Source Verified  Dec-2018   

   loan_status pymnt_plan  url  desc             purpose  \
0  Charged Off          n  NaN   NaN  debt_consolidation   
1      Current          n  NaN   NaN               other   
2   Fully Paid          n  NaN   NaN         credit_card   
3      Current          n  NaN   NaN         credit_card   
4  Charged Off          n  NaN   NaN  debt_consolidation   

                     title zip_code addr_state    dti  delinq_2yrs  \
0       Debt consolidation    681xx         NE  30.65          1.0   
1                    Other    917xx         CA  17.94          0.0   
2  Credit card refinancing    020xx         MA  12.37          0.0   
3  Credit card refinancing    660xx         KS  14.90          0.0   
4       Debt consolidation    920xx         CA  37.24          0.0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0         Jun-1979             2.0                    15.0   
1         Nov-2001             1.0                     NaN   
2         Oct-1997             0.0                     NaN   
3         May-2007             0.0                     NaN   
4         Oct-2007             0.0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN       6.0      0.0    15048.0        73%       22.0   
1                     NaN      24.0      0.0    43721.0      59.2%       29.0   
2                     NaN      21.0      0.0    36812.0      65.7%       37.0   
3                   112.0       7.0      1.0    10677.0      54.2%       12.0   
4                    86.0      15.0      1.0    18890.0      39.8%       36.0   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w       0.00           0.00  8169.130000          8169.13   
1                   w    9726.53        9726.53  8788.870000          8788.87   
2                   w       0.00           0.00  8735.148975          8735.15   
3                   w    7787.21        7787.21  3805.020000          3805.02   
4                   w       0.00           0.00  2645.410000          2645.41   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          5731.54        2437.59                 0.0         0.0   
1          6273.47        2515.40                 0.0         0.0   
2          8425.00         310.15                 0.0         0.0   
3          2212.79        1592.23                 0.0         0.0   
4           754.24        1891.17                 0.0         0.0   

   collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d  \
0                      0.0     Feb-2020       

In [0]:
# The extra rows at the top and bottom of the file have done two things:
#1) The top row has made it so that the entire dataset is being intrepreted as column header
#2) The bottom two rows have been read into the 'id' column 
#and are causing every column to have at least the NaN walues in it. 

In [0]:
# We can fix the header problem by using the 'skiprows' parameter

df = pd.read_csv('LoanStats_2018Q4.csv', skiprows=1)
 
 #Let's look at tnhe shape and head of the data.
 #We are doing this to get an idea of its shape and size.
 
print(df.shape)
df.head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(128414, 144)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  NaN        NaN    20000.0      20000.0          20000.0   36 months   
1  NaN        NaN    16000.0      16000.0          16000.0   36 months   
2  NaN        NaN     8425.0       8425.0           8425.0   36 months   
3  NaN        NaN    10000.0      10000.0          10000.0   60 months   
4  NaN        NaN    30000.0      30000.0          30000.0   60 months   

  int_rate  installment grade sub_grade                        emp_title  \
0   14.47%       688.13     C        C2                       bus driver   
1   14.47%       550.51     C        C2                        Carpenter   
2   27.27%       345.18     E        E5  Senior Director Risk Management   
3   12.98%       227.43     B        B5                          Printer   
4   26.31%       903.73     E        E4                      Casino Host   

  emp_length home_ownership  annual_inc verification_status   issue_d  \
0    4 years            OWN     52000.0     Source Verified  Dec-2018   
1  10+ years           RENT    150000.0     Source Verified  Dec-2018   
2    3 years       MORTGAGE    450000.0            Verified  Dec-2018   
3    9 years           RENT     60000.0        Not Verified  Dec-2018   
4  10+ years            OWN     62000.0     Source Verified  Dec-2018   

   loan_status pymnt_plan  url  desc             purpose  \
0  Charged Off          n  NaN   NaN  debt_consolidation   
1      Current          n  NaN   NaN               other   
2   Fully Paid          n  NaN   NaN         credit_card   
3      Current          n  NaN   NaN         credit_card   
4  Charged Off          n  NaN   NaN  debt_consolidation   

                     title zip_code addr_state    dti  delinq_2yrs  \
0       Debt consolidation    681xx         NE  30.65          1.0   
1                    Other    917xx         CA  17.94          0.0   
2  Credit card refinancing    020xx         MA  12.37          0.0   
3  Credit card refinancing    660xx         KS  14.90          0.0   
4       Debt consolidation    920xx         CA  37.24          0.0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0         Jun-1979             2.0                    15.0   
1         Nov-2001             1.0                     NaN   
2         Oct-1997             0.0                     NaN   
3         May-2007             0.0                     NaN   
4         Oct-2007             0.0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN       6.0      0.0    15048.0        73%       22.0   
1                     NaN      24.0      0.0    43721.0      59.2%       29.0   
2                     NaN      21.0      0.0    36812.0      65.7%       37.0   
3                   112.0       7.0      1.0    10677.0      54.2%       12.0   
4                    86.0      15.0      1.0    18890.0      39.8%       36.0   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w       0.00           0.00  8169.130000          8169.13   
1                   w    9726.53        9726.53  8788.870000          8788.87   
2                   w       0.00           0.00  8735.148975          8735.15   
3                   w    7787.21        7787.21  3805.020000          3805.02   
4                   w       0.00           0.00  2645.410000          2645.41   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          5731.54        2437.59                 0.0         0.0   
1          6273.47        2515.40                 0.0         0.0   
2          8425.00         310.15                 0.0         0.0   
3          2212.79        1592.23                 0.0         0.0   
4           754.24        1891.17                 0.0         0.0   

   collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d  \
0                      0.0     Feb-2020       

#Look at the Nan values of each column so that you can see the problem that the extra rows are the bottom of the file are creating for us. 


'Total amount funded in policy code 1: 2050909275
Total amount funded in policy code 2: 820109297'

#Those two lines of code at the beginning and the end of the dataset are causing problems when we read the CSV.

In [0]:
#Sum null values by column and sort from the least to greatest value.

df.isnull().sum().sort_values()

loan_amnt                                          0
mo_sin_rcnt_rev_tl_op                              0
mo_sin_old_rev_tl_op                               0
delinq_amnt                                        0
chargeoff_within_12_mths                           0
acc_open_past_24mths                               0
total_cu_tl                                        0
inq_fi                                             0
total_rev_hi_lim                                   0
max_bal_bc                                         0
open_rv_24m                                        0
open_rv_12m                                        0
total_bal_il                                       0
open_il_24m                                        0
open_il_12m                                        0
open_act_il                                        0
open_acc_6m                                        0
tot_cur_bal                                        0
tot_coll_amt                                  

In [0]:
# Address the extra NaNs in each column by skipping the footer as well. 

df.tail()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  \
128407      23000        23000          23000.0   36 months   15.02%   
128408      10000        10000          10000.0   36 months   15.02%   
128409       5000         5000           5000.0   36 months   13.56%   
128410      10000        10000           9750.0   36 months   11.06%   
128411      10000        10000          10000.0   36 months   16.91%   

        installment grade sub_grade             emp_title emp_length  \
128407       797.53     C        C3        Tax Consultant  10+ years   
128408       346.76     C        C3        security guard    5 years   
128409       169.83     C        C1          Payoff Clerk  10+ years   
128410       327.68     B        B3                   NaN        NaN   
128411       356.08     C        C5  Key Accounts Manager    2 years   

       home_ownership  annual_inc verification_status   issue_d  loan_status  \
128407       MORTGAGE     75000.0     Source Verified  Oct-2018  Charged Off   
128408       MORTGAGE     38000.0        Not Verified  Oct-2018      Current   
128409       MORTGAGE     35360.0        Not Verified  Oct-2018      Current   
128410           RENT     44400.0     Source Verified  Oct-2018      Current   
128411           RENT     80000.0        Not Verified  Oct-2018      Current   

       pymnt_plan             purpose                    title zip_code  \
128407          n  debt_consolidation       Debt consolidation    352xx   
128408          n  debt_consolidation       Debt consolidation    443xx   
128409          n  debt_consolidation       Debt consolidation    381xx   
128410          n         credit_card  Credit card refinancing    980xx   
128411          n               other                    Other    021xx   

       addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
128407         AL  20.95            1         Aug-1985               2   
128408         OH  13.16            3         Jul-1982               0   
128409         TN  11.30            1         Jun-2006               0   
128410         WA  11.78            0         Oct-2008               2   
128411         MA  17.72            1         Sep-2006               0   

        mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
128407                    22.0                     NaN        12        0   
128408                     6.0                     NaN        11        0   
128409                    21.0                     NaN         9        0   
128410                    40.0                     NaN        15        0   
128411                    14.0                     NaN        17        0   

        revol_bal revol_util  total_acc initial_list_status  out_prncp  \
128407      22465      43.6%         28                   w       0.00   
128408       5634      37.1%         16                   w    5280.05   
128409       2597      27.3%         15                   f    2613.06   
128410       6269      13.1%         25                   f    5133.94   
128411       1942      30.8%         31                   w    5349.59   

        out_prncp_inv  total_pymnt  total_pymnt_inv  total_rec_prncp  \
128407           0.00      1547.08          1547.08          1025.67   
128408        5280.05      6580.10          6580.10          4719.95   
128409        2613.06      3223.00          3223.00          2386.94   
128410        5005.59      6210.56          6055.30          4866.06   
128411        5349.59      6756.13          6756.13          4650.41   

        total_rec_int  total_rec_late_fee  recoveries  \
128407         521.41                 0.0         0.0   
128408        1860.15                 0.0         0.0   
128409         836.06                 0.0         0.0   
128410        1344.50                 0.0         0.0   
128411        2105.72                 0.0         0.0   

        collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d  \
128

#This line of code will remove the one row in the dataset at the beginning and the last two rows at the end. Removing these will help solve the issue with the trouble we were having when we read the CSV.

In [0]:
df = pd.read_csv('LoanStats_2018Q4.csv', skiprows=1, skipfooter=2, engine='python')
# Be sure to include the engine being used to process the edit in the CSV data. 

#Lets check our work.
print(df.shape)
df.head()

(128412, 144)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0 NaN        NaN      20000        20000          20000.0   36 months   
1 NaN        NaN      16000        16000          16000.0   36 months   
2 NaN        NaN       8425         8425           8425.0   36 months   
3 NaN        NaN      10000        10000          10000.0   60 months   
4 NaN        NaN      30000        30000          30000.0   60 months   

  int_rate  installment grade sub_grade                        emp_title  \
0   14.47%       688.13     C        C2                       bus driver   
1   14.47%       550.51     C        C2                        Carpenter   
2   27.27%       345.18     E        E5  Senior Director Risk Management   
3   12.98%       227.43     B        B5                          Printer   
4   26.31%       903.73     E        E4                      Casino Host   

  emp_length home_ownership  annual_inc verification_status   issue_d  \
0    4 years            OWN     52000.0     Source Verified  Dec-2018   
1  10+ years           RENT    150000.0     Source Verified  Dec-2018   
2    3 years       MORTGAGE    450000.0            Verified  Dec-2018   
3    9 years           RENT     60000.0        Not Verified  Dec-2018   
4  10+ years            OWN     62000.0     Source Verified  Dec-2018   

   loan_status pymnt_plan  url  desc             purpose  \
0  Charged Off          n  NaN   NaN  debt_consolidation   
1      Current          n  NaN   NaN               other   
2   Fully Paid          n  NaN   NaN         credit_card   
3      Current          n  NaN   NaN         credit_card   
4  Charged Off          n  NaN   NaN  debt_consolidation   

                     title zip_code addr_state    dti  delinq_2yrs  \
0       Debt consolidation    681xx         NE  30.65            1   
1                    Other    917xx         CA  17.94            0   
2  Credit card refinancing    020xx         MA  12.37            0   
3  Credit card refinancing    660xx         KS  14.90            0   
4       Debt consolidation    920xx         CA  37.24            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0         Jun-1979               2                    15.0   
1         Nov-2001               1                     NaN   
2         Oct-1997               0                     NaN   
3         May-2007               0                     NaN   
4         Oct-2007               0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0      15048        73%         22   
1                     NaN        24        0      43721      59.2%         29   
2                     NaN        21        0      36812      65.7%         37   
3                   112.0         7        1      10677      54.2%         12   
4                    86.0        15        1      18890      39.8%         36   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w       0.00           0.00  8169.130000          8169.13   
1                   w    9726.53        9726.53  8788.870000          8788.87   
2                   w       0.00           0.00  8735.148975          8735.15   
3                   w    7787.21        7787.21  3805.020000          3805.02   
4                   w       0.00           0.00  2645.410000          2645.41   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          5731.54        2437.59                 0.0         0.0   
1          6273.47        2515.40                 0.0         0.0   
2          8425.00         310.15                 0.0         0.0   
3          2212.79        1592.23                 0.0         0.0   
4           754.24        1891.17                 0.0         0.0   

   collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d  \
0                      0.0     Feb-2020            

In [0]:
# df.isnull().sum() will show the NaN values.
# .sort_values() will organize to ensure a change has been made


df.isnull().sum().sort_values()

inq_fi                                             0
mo_sin_old_rev_tl_op                               0
delinq_amnt                                        0
chargeoff_within_12_mths                           0
acc_open_past_24mths                               0
inq_last_12m                                       0
total_cu_tl                                        0
total_rev_hi_lim                                   0
open_rv_24m                                        0
open_rv_12m                                        0
total_bal_il                                       0
open_il_24m                                        0
open_il_12m                                        0
open_act_il                                        0
open_acc_6m                                        0
tot_cur_bal                                        0
tot_coll_amt                                       0
acc_now_delinq                                     0
application_type                              

(Above) You will see that after removing the first and last two rows which had the following information:

"Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)" (1st row)

"Total amount funded in policy code 1: 2050909275
Total amount funded in policy code 2: 820109297" (Last two rows)

... that those rows contribute to the NaN values. Now the values are showing "0" indicating that the problem has been resolved. 






After futher inverstigation, you will notice that there are columns completely made up of NaN values. Why might LendingClub have included columns in their dataset that are 100% blank.

desc                                          128412

url                                           128412

member_id                                     128412

id                                            128412

In [0]:
#In order to drop columns, you will need to use the .drop() method. 

df =  df.drop(['desc', 'url', 'member_id', 'id'], axis=1)

In [0]:
#Let's check a few rows from the dataset to ensure the changes were made
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0      20000        20000          20000.0   36 months   14.47%       688.13   
1      16000        16000          16000.0   36 months   14.47%       550.51   
2       8425         8425           8425.0   36 months   27.27%       345.18   
3      10000        10000          10000.0   60 months   12.98%       227.43   
4      30000        30000          30000.0   60 months   26.31%       903.73   

  grade sub_grade                        emp_title emp_length home_ownership  \
0     C        C2                       bus driver    4 years            OWN   
1     C        C2                        Carpenter  10+ years           RENT   
2     E        E5  Senior Director Risk Management    3 years       MORTGAGE   
3     B        B5                          Printer    9 years           RENT   
4     E        E4                      Casino Host  10+ years            OWN   

   annual_inc verification_status   issue_d  loan_status pymnt_plan  \
0     52000.0     Source Verified  Dec-2018  Charged Off          n   
1    150000.0     Source Verified  Dec-2018      Current          n   
2    450000.0            Verified  Dec-2018   Fully Paid          n   
3     60000.0        Not Verified  Dec-2018      Current          n   
4     62000.0     Source Verified  Dec-2018  Charged Off          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    681xx         NE  30.65   
1               other                    Other    917xx         CA  17.94   
2         credit_card  Credit card refinancing    020xx         MA  12.37   
3         credit_card  Credit card refinancing    660xx         KS  14.90   
4  debt_consolidation       Debt consolidation    920xx         CA  37.24   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            1         Jun-1979               2                    15.0   
1            0         Nov-2001               1                     NaN   
2            0         Oct-1997               0                     NaN   
3            0         May-2007               0                     NaN   
4            0         Oct-2007               0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0      15048        73%         22   
1                     NaN        24        0      43721      59.2%         29   
2                     NaN        21        0      36812      65.7%         37   
3                   112.0         7        1      10677      54.2%         12   
4                    86.0        15        1      18890      39.8%         36   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w       0.00           0.00  8169.130000          8169.13   
1                   w    9726.53        9726.53  8788.870000          8788.87   
2                   w       0.00           0.00  8735.148975          8735.15   
3                   w    7787.21        7787.21  3805.020000          3805.02   
4                   w       0.00           0.00  2645.410000          2645.41   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          5731.54        2437.59                 0.0         0.0   
1          6273.47        2515.40                 0.0         0.0   
2          8425.00         310.15                 0.0         0.0   
3          2212.79        1592.23                 0.0         0.0   
4           754.24        1891.17                 0.0         0.0   

   collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d  \
0                      0.0     Feb-2020             0.00          NaN   
1                      0.0     May-2020             0.00     Jun-2020   
2                      0.0     Feb-2019          8754.30          NaN   
3                  

Looks good to me! (:

Now, let's look at the types (.dtypes)

In [0]:
df.dtypes

loan_amnt                                       int64
funded_amnt                                     int64
funded_amnt_inv                               float64
term                                           object
int_rate                                       object
installment                                   float64
grade                                          object
sub_grade                                      object
emp_title                                      object
emp_length                                     object
home_ownership                                 object
annual_inc                                    float64
verification_status                            object
issue_d                                        object
loan_status                                    object
pymnt_plan                                     object
purpose                                        object
title                                          object
zip_code                    

When preparing a dataset for machine learning model we typically dont want to leave any string values in our dataset -- becuase it's hard to do math on words. 

In [0]:
#Lets Look at the first 10 values of the int_rate column. 

df['int_rate'].head(10)

0     14.47%
1     14.47%
2     27.27%
3     12.98%
4     26.31%
5     17.97%
6     11.80%
7     10.72%
8     16.14%
9     12.98%
Name: int_rate, dtype: object

In [0]:
#Look at a specific value from the int_rate column

df['int_rate'][3]

' 12.98%'

In [0]:
#An alternative to get a slice of dataframe

df.iloc[0:10, 0:2]

loan_amnt  funded_amnt
0      20000        20000
1      16000        16000
2       8425         8425
3      10000        10000
4      30000        30000
5      13000        13000
6       6500         6500
7       7500         7500
8      14000        14000
9      25000        25000

These are the problems that we need to address with this column:

-the string column that should be numeric
' 12.98%'

-Percent sign % included with the number needs to be removed

-Leading space at the beginning of the string needs to be removed.

In [0]:
#Lets start with just fixing a single string. 
#If we can fix one, we can usually fix all of them

int_rate = ' 12.98%'

In [0]:
#Removes the leading space in the string. 
#Not the string with the ' '

int_rate.strip()

'12.98%'

In [0]:
#Removes the '%' sign in the string

int_rate.strip('%')

' 12.98'

In [0]:
#The proccess of added two dot methods is called "chaining"
int_rate.strip().strip('%')

'12.98'

In [0]:
#Let's take a second to see what type our column is.
type(int_rate.strip().strip('%'))

str

In [0]:
#"Cast" the string value to a float.
#"Casting" 

float(int_rate.strip().strip('%'))

12.98

In [0]:
#Let's check the tyoe again...

type(float(int_rate.strip().strip('%')))

float

Let's write a function to make our solution reusable for other cells in a column.

In [0]:
# Write a function that can do what we have written above to any
# string that is passed to it.

def int_rate_to_float(cell_contents):
  return float(cell_contents.strip().strip('%'))

In [0]:
#Test out the function by calling it on our example. 
# Let's Try.

int_rate_to_float(int_rate)

12.98

In [0]:
#Is the data type correct. Verify.

type(int_rate_to_float(int_rate))

float

In [0]:
cleaned = []

for int_rate in df['int_rate']:
  cleaned.append(int_rate_to_float(int_rate))

  df['int_rate_cleaned'] = pd.Series(cleaned)

KeyboardInterrupt: ignored

In [0]:
#Let's Check our work
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0      20000        20000          20000.0   36 months   14.47%       688.13   
1      16000        16000          16000.0   36 months   14.47%       550.51   
2       8425         8425           8425.0   36 months   27.27%       345.18   
3      10000        10000          10000.0   60 months   12.98%       227.43   
4      30000        30000          30000.0   60 months   26.31%       903.73   

  grade sub_grade                        emp_title emp_length home_ownership  \
0     C        C2                       bus driver    4 years            OWN   
1     C        C2                        Carpenter  10+ years           RENT   
2     E        E5  Senior Director Risk Management    3 years       MORTGAGE   
3     B        B5                          Printer    9 years           RENT   
4     E        E4                      Casino Host  10+ years            OWN   

   annual_inc verification_status   issue_d  loan_status pymnt_plan  \
0     52000.0     Source Verified  Dec-2018  Charged Off          n   
1    150000.0     Source Verified  Dec-2018      Current          n   
2    450000.0            Verified  Dec-2018   Fully Paid          n   
3     60000.0        Not Verified  Dec-2018      Current          n   
4     62000.0     Source Verified  Dec-2018  Charged Off          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    681xx         NE  30.65   
1               other                    Other    917xx         CA  17.94   
2         credit_card  Credit card refinancing    020xx         MA  12.37   
3         credit_card  Credit card refinancing    660xx         KS  14.90   
4  debt_consolidation       Debt consolidation    920xx         CA  37.24   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            1         Jun-1979               2                    15.0   
1            0         Nov-2001               1                     NaN   
2            0         Oct-1997               0                     NaN   
3            0         May-2007               0                     NaN   
4            0         Oct-2007               0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0      15048        73%         22   
1                     NaN        24        0      43721      59.2%         29   
2                     NaN        21        0      36812      65.7%         37   
3                   112.0         7        1      10677      54.2%         12   
4                    86.0        15        1      18890      39.8%         36   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w       0.00           0.00  8169.130000          8169.13   
1                   w    9726.53        9726.53  8788.870000          8788.87   
2                   w       0.00           0.00  8735.148975          8735.15   
3                   w    7787.21        7787.21  3805.020000          3805.02   
4                   w       0.00           0.00  2645.410000          2645.41   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          5731.54        2437.59                 0.0         0.0   
1          6273.47        2515.40                 0.0         0.0   
2          8425.00         310.15                 0.0         0.0   
3          2212.79        1592.23                 0.0         0.0   
4           754.24        1891.17                 0.0         0.0   

   collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d  \
0                      0.0     Feb-2020             0.00          NaN   
1                      0.0     May-2020             0.00     Jun-2020   
2                      0.0     Feb-2019          8754.30          NaN   
3                  

In [0]:
# What type of data is held in our new columns?
df.dtypes[-5:]

#Let's look at the datatypes of the last 5 columns.

settlement_date           object
settlement_amount        float64
settlement_percentage    float64
settlement_term          float64
int_rate_cleaned         float64
dtype: object

In [0]:
int_rates = [' 16.57%', ' 16.52%', ' 16.58%', ' 16.41%', ' 15.47%']

[int_rate_to_float(item) for item in int_rates]

[16.57, 16.52, 16.58, 16.41, 15.47]

In [0]:
df['int_rate'].values

array([' 14.47%', ' 14.47%', ' 27.27%', ..., ' 13.56%', ' 11.06%',
       ' 16.91%'], dtype=object)

In [0]:
# use .apply() to fix the int_rate column

df['int_rate'] = df['int_rate'].apply(int_rate_to_float)

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment  \
0      20000        20000          20000.0   36 months     14.47       688.13   
1      16000        16000          16000.0   36 months     14.47       550.51   
2       8425         8425           8425.0   36 months     27.27       345.18   
3      10000        10000          10000.0   60 months     12.98       227.43   
4      30000        30000          30000.0   60 months     26.31       903.73   

  grade sub_grade                        emp_title emp_length home_ownership  \
0     C        C2                       bus driver    4 years            OWN   
1     C        C2                        Carpenter  10+ years           RENT   
2     E        E5  Senior Director Risk Management    3 years       MORTGAGE   
3     B        B5                          Printer    9 years           RENT   
4     E        E4                      Casino Host  10+ years            OWN   

   annual_inc verification_status   issue_d  loan_status pymnt_plan  \
0     52000.0     Source Verified  Dec-2018  Charged Off          n   
1    150000.0     Source Verified  Dec-2018      Current          n   
2    450000.0            Verified  Dec-2018   Fully Paid          n   
3     60000.0        Not Verified  Dec-2018      Current          n   
4     62000.0     Source Verified  Dec-2018  Charged Off          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    681xx         NE  30.65   
1               other                    Other    917xx         CA  17.94   
2         credit_card  Credit card refinancing    020xx         MA  12.37   
3         credit_card  Credit card refinancing    660xx         KS  14.90   
4  debt_consolidation       Debt consolidation    920xx         CA  37.24   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            1         Jun-1979               2                    15.0   
1            0         Nov-2001               1                     NaN   
2            0         Oct-1997               0                     NaN   
3            0         May-2007               0                     NaN   
4            0         Oct-2007               0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0      15048        73%         22   
1                     NaN        24        0      43721      59.2%         29   
2                     NaN        21        0      36812      65.7%         37   
3                   112.0         7        1      10677      54.2%         12   
4                    86.0        15        1      18890      39.8%         36   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w       0.00           0.00  8169.130000          8169.13   
1                   w    9726.53        9726.53  8788.870000          8788.87   
2                   w       0.00           0.00  8735.148975          8735.15   
3                   w    7787.21        7787.21  3805.020000          3805.02   
4                   w       0.00           0.00  2645.410000          2645.41   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          5731.54        2437.59                 0.0         0.0   
1          6273.47        2515.40                 0.0         0.0   
2          8425.00         310.15                 0.0         0.0   
3          2212.79        1592.23                 0.0         0.0   
4           754.24        1891.17                 0.0         0.0   

   collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d  \
0                      0.0     Feb-2020             0.00          NaN   
1                      0.0     May-2020             0.00     Jun-2020   
2                      0.0     Feb-2019          8754.30          NaN   
3            

In [0]:
# Look at the top 20 employment titles
df['emp_title'].value_counts(dropna=False).head(20)

NaN                   20947
Teacher                2090
Manager                1773
Registered Nurse        952
Driver                  924
RN                      726
Supervisor              697
Sales                   580
Project Manager         526
General Manager         523
Office Manager          521
Owner                   420
Director                402
Operations Manager      387
Truck Driver            387
Nurse                   326
Engineer                325
Sales Manager           304
manager                 301
Supervisor              270
Name: emp_title, dtype: int64

In [0]:
# How often is the employment_title null?
df['emp_title'].isnull().sum()

20947

In [0]:
# Create some examples that represent the cases that we want to clean up

import numpy as np

examples = ['owner', 'Supervisor', ' Project Manager', np.NaN]

In [0]:
# Write a function to clean up these use cases and increase uniformity.

def clean_emp_title(title):
  if isinstance(title, str):
    return title.strip().title()
  else:
    return "Unknown"

for example in examples:
  print(clean_emp_title(example))

Owner
Supervisor
Project Manager
Unknown


In [0]:
# list comprehensions can combine function calls and for loops over lists
# into one succinct and fairly readable single line of code.

df['new_column'] = [clean_emp_title(title) for title in examples]

ValueError: ignored

In [0]:
# We have a function that works as expected. Lets apply it to our column.
# This time we'll overwrite the original column

df['emp_title'] = df['emp_title'].apply(clean_emp_title)

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment  \
0      20000        20000          20000.0   36 months     14.47       688.13   
1      16000        16000          16000.0   36 months     14.47       550.51   
2       8425         8425           8425.0   36 months     27.27       345.18   
3      10000        10000          10000.0   60 months     12.98       227.43   
4      30000        30000          30000.0   60 months     26.31       903.73   

  grade sub_grade                        emp_title emp_length home_ownership  \
0     C        C2                       Bus Driver    4 years            OWN   
1     C        C2                        Carpenter  10+ years           RENT   
2     E        E5  Senior Director Risk Management    3 years       MORTGAGE   
3     B        B5                          Printer    9 years           RENT   
4     E        E4                      Casino Host  10+ years            OWN   

   annual_inc verification_status   issue_d  loan_status pymnt_plan  \
0     52000.0     Source Verified  Dec-2018  Charged Off          n   
1    150000.0     Source Verified  Dec-2018      Current          n   
2    450000.0            Verified  Dec-2018   Fully Paid          n   
3     60000.0        Not Verified  Dec-2018      Current          n   
4     62000.0     Source Verified  Dec-2018  Charged Off          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    681xx         NE  30.65   
1               other                    Other    917xx         CA  17.94   
2         credit_card  Credit card refinancing    020xx         MA  12.37   
3         credit_card  Credit card refinancing    660xx         KS  14.90   
4  debt_consolidation       Debt consolidation    920xx         CA  37.24   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            1         Jun-1979               2                    15.0   
1            0         Nov-2001               1                     NaN   
2            0         Oct-1997               0                     NaN   
3            0         May-2007               0                     NaN   
4            0         Oct-2007               0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0      15048        73%         22   
1                     NaN        24        0      43721      59.2%         29   
2                     NaN        21        0      36812      65.7%         37   
3                   112.0         7        1      10677      54.2%         12   
4                    86.0        15        1      18890      39.8%         36   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w       0.00           0.00  8169.130000          8169.13   
1                   w    9726.53        9726.53  8788.870000          8788.87   
2                   w       0.00           0.00  8735.148975          8735.15   
3                   w    7787.21        7787.21  3805.020000          3805.02   
4                   w       0.00           0.00  2645.410000          2645.41   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          5731.54        2437.59                 0.0         0.0   
1          6273.47        2515.40                 0.0         0.0   
2          8425.00         310.15                 0.0         0.0   
3          2212.79        1592.23                 0.0         0.0   
4           754.24        1891.17                 0.0         0.0   

   collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d  \
0                      0.0     Feb-2020             0.00          NaN   
1                      0.0     May-2020             0.00     Jun-2020   
2                      0.0     Feb-2019          8754.30          NaN   
3            

In [0]:
# Look at the top 20 employment titles

df['emp_title'].value_counts(dropna=False).head(20)

Unknown               20947
Teacher                2557
Manager                2395
Registered Nurse       1418
Driver                 1258
Supervisor             1160
Truck Driver            920
Rn                      834
Office Manager          805
Sales                   803
General Manager         791
Project Manager         720
Owner                   625
Director                523
Operations Manager      518
Sales Manager           500
Police Officer          440
Nurse                   425
Technician              420
Engineer                412
Name: emp_title, dtype: int64

In [0]:
# How many different unique employment titles are there currently?
df['emp_title'].nunique()

34902

In [0]:
# How often is the employment_title null (NaN)?
df['emp_title'].isnull().sum()

0

Working with Dates and Times with Pandas

In [0]:
[col for col in df if col.endswith('_d')]

['issue_d', 'last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d']

In [0]:
df['issue_d'].head()

0    Dec-2018
1    Dec-2018
2    Dec-2018
3    Dec-2018
4    Dec-2018
Name: issue_d, dtype: object

In [0]:
df['issue_d'].tail()

128407    Oct-2018
128408    Oct-2018
128409    Oct-2018
128410    Oct-2018
128411    Oct-2018
Name: issue_d, dtype: object

In [0]:
df['issue_d'] = pd.to_datetime(df['issue_d'], infer_datetime_format=True)

In [0]:
df['issue_d'].head()

0   2018-12-01
1   2018-12-01
2   2018-12-01
3   2018-12-01
4   2018-12-01
Name: issue_d, dtype: datetime64[ns]

In [0]:
df['issue_d'][0]

Timestamp('2018-12-01 00:00:00')

In [0]:
df['issue_d'].dt.year

0         2018
1         2018
2         2018
3         2018
4         2018
          ... 
128407    2018
128408    2018
128409    2018
128410    2018
128411    2018
Name: issue_d, Length: 128412, dtype: int64

In [0]:
df['issue_d'].dt.month

0         12
1         12
2         12
3         12
4         12
          ..
128407    10
128408    10
128409    10
128410    10
128411    10
Name: issue_d, Length: 128412, dtype: int64

In [0]:
df['issue_year'] = df['issue_d'].dt.year
df['issue_month'] = df['issue_d'].dt.month

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment  \
0      20000        20000          20000.0   36 months     14.47       688.13   
1      16000        16000          16000.0   36 months     14.47       550.51   
2       8425         8425           8425.0   36 months     27.27       345.18   
3      10000        10000          10000.0   60 months     12.98       227.43   
4      30000        30000          30000.0   60 months     26.31       903.73   

  grade sub_grade                        emp_title emp_length home_ownership  \
0     C        C2                       Bus Driver    4 years            OWN   
1     C        C2                        Carpenter  10+ years           RENT   
2     E        E5  Senior Director Risk Management    3 years       MORTGAGE   
3     B        B5                          Printer    9 years           RENT   
4     E        E4                      Casino Host  10+ years            OWN   

   annual_inc verification_status    issue_d  loan_status pymnt_plan  \
0     52000.0     Source Verified 2018-12-01  Charged Off          n   
1    150000.0     Source Verified 2018-12-01      Current          n   
2    450000.0            Verified 2018-12-01   Fully Paid          n   
3     60000.0        Not Verified 2018-12-01      Current          n   
4     62000.0     Source Verified 2018-12-01  Charged Off          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    681xx         NE  30.65   
1               other                    Other    917xx         CA  17.94   
2         credit_card  Credit card refinancing    020xx         MA  12.37   
3         credit_card  Credit card refinancing    660xx         KS  14.90   
4  debt_consolidation       Debt consolidation    920xx         CA  37.24   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            1         Jun-1979               2                    15.0   
1            0         Nov-2001               1                     NaN   
2            0         Oct-1997               0                     NaN   
3            0         May-2007               0                     NaN   
4            0         Oct-2007               0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0      15048        73%         22   
1                     NaN        24        0      43721      59.2%         29   
2                     NaN        21        0      36812      65.7%         37   
3                   112.0         7        1      10677      54.2%         12   
4                    86.0        15        1      18890      39.8%         36   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w       0.00           0.00  8169.130000          8169.13   
1                   w    9726.53        9726.53  8788.870000          8788.87   
2                   w       0.00           0.00  8735.148975          8735.15   
3                   w    7787.21        7787.21  3805.020000          3805.02   
4                   w       0.00           0.00  2645.410000          2645.41   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          5731.54        2437.59                 0.0         0.0   
1          6273.47        2515.40                 0.0         0.0   
2          8425.00         310.15                 0.0         0.0   
3          2212.79        1592.23                 0.0         0.0   
4           754.24        1891.17                 0.0         0.0   

   collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d  \
0                      0.0     Feb-2020             0.00          NaN   
1                      0.0     May-2020             0.00     Jun-2020   
2                      0.0     Feb-2019          8754.30          NaN   
3      

In [0]:
df.dtypes

loan_amnt                                              int64
funded_amnt                                            int64
funded_amnt_inv                                      float64
term                                                  object
int_rate                                             float64
installment                                          float64
grade                                                 object
sub_grade                                             object
emp_title                                             object
emp_length                                            object
home_ownership                                        object
annual_inc                                           float64
verification_status                                   object
issue_d                                       datetime64[ns]
loan_status                                           object
pymnt_plan                                            object
purpose                 

In [0]:
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'], infer_datetime_format=True)

In [0]:
df['earliest_cr_line'].head()

0   1979-06-01
1   2001-11-01
2   1997-10-01
3   2007-05-01
4   2007-10-01
Name: earliest_cr_line, dtype: datetime64[ns]

In [0]:
(df['issue_d'] - df['earliest_cr_line']).dt.days / 365

0         39.528767
1         17.093151
2         21.180822
3         11.594521
4         11.175342
            ...    
128407    33.189041
128408    36.276712
128409    12.342466
128410    10.005479
128411    12.090411
Length: 128412, dtype: float64

In [0]:
df['length_of_credit_history'] = (df['issue_d'] - df['earliest_cr_line']).dt.days / 365

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment  \
0      20000        20000          20000.0   36 months     14.47       688.13   
1      16000        16000          16000.0   36 months     14.47       550.51   
2       8425         8425           8425.0   36 months     27.27       345.18   
3      10000        10000          10000.0   60 months     12.98       227.43   
4      30000        30000          30000.0   60 months     26.31       903.73   

  grade sub_grade                        emp_title emp_length home_ownership  \
0     C        C2                       Bus Driver    4 years            OWN   
1     C        C2                        Carpenter  10+ years           RENT   
2     E        E5  Senior Director Risk Management    3 years       MORTGAGE   
3     B        B5                          Printer    9 years           RENT   
4     E        E4                      Casino Host  10+ years            OWN   

   annual_inc verification_status    issue_d  loan_status pymnt_plan  \
0     52000.0     Source Verified 2018-12-01  Charged Off          n   
1    150000.0     Source Verified 2018-12-01      Current          n   
2    450000.0            Verified 2018-12-01   Fully Paid          n   
3     60000.0        Not Verified 2018-12-01      Current          n   
4     62000.0     Source Verified 2018-12-01  Charged Off          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    681xx         NE  30.65   
1               other                    Other    917xx         CA  17.94   
2         credit_card  Credit card refinancing    020xx         MA  12.37   
3         credit_card  Credit card refinancing    660xx         KS  14.90   
4  debt_consolidation       Debt consolidation    920xx         CA  37.24   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            1       1979-06-01               2                    15.0   
1            0       2001-11-01               1                     NaN   
2            0       1997-10-01               0                     NaN   
3            0       2007-05-01               0                     NaN   
4            0       2007-10-01               0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0      15048        73%         22   
1                     NaN        24        0      43721      59.2%         29   
2                     NaN        21        0      36812      65.7%         37   
3                   112.0         7        1      10677      54.2%         12   
4                    86.0        15        1      18890      39.8%         36   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w       0.00           0.00  8169.130000          8169.13   
1                   w    9726.53        9726.53  8788.870000          8788.87   
2                   w       0.00           0.00  8735.148975          8735.15   
3                   w    7787.21        7787.21  3805.020000          3805.02   
4                   w       0.00           0.00  2645.410000          2645.41   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          5731.54        2437.59                 0.0         0.0   
1          6273.47        2515.40                 0.0         0.0   
2          8425.00         310.15                 0.0         0.0   
3          2212.79        1592.23                 0.0         0.0   
4           754.24        1891.17                 0.0         0.0   

   collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d  \
0                      0.0     Feb-2020             0.00          NaN   
1                      0.0     May-2020             0.00     Jun-2020   
2                      0.0     Feb-2019          8754.30          NaN   
3      

What is the oldest credit history that was involced in Q4 2018?

In [0]:
#In order to get the max number for the oldest credit history from Q4 in 2018
# You will need to use the .describe() method. 
# Use the max


df['length_of_credit_history'].describe()

count    128412.000000
mean         16.054497
std           7.908317
min           3.084932
25%          11.093151
50%          14.427397
75%          19.846575
max          68.961644
Name: length_of_credit_history, dtype: float64

Converting the Term column to an Integer, INT.

In [0]:
#Look at the first 10 values of the term column from the Lending Club CSV.
# Use the .head()
# Remeber Zero index rule. **
df['term'].head(10)

0     36 months
1     36 months
2     36 months
3     60 months
4     60 months
5     36 months
6     36 months
7     36 months
8     36 months
9     60 months
Name: term, dtype: object

In [0]:
#Look at a specific value from the term column

df['term'][3]

' 60 months'

In [0]:
#Look at another specific value fromm the term column

df['term'][5]

' 36 months'

In [0]:
#Let's start to fix a single string
term = ' 60 months'

In [0]:
term.strip()
type(term.strip())

#Remove the leading space
#Check the type

str

In [0]:
int(term.strip().strip(' months'))

# .strip() will 'strip anything provide within the (). 
# For ex: look at how the ' months' with the space was removed from the STR.
# Chained and changed from STR to INT by casting.
# int(------------) "Cast" the STR value to an INT. 

60

In [0]:
# Write a function that can do what we have written above to any 
# string that is passsed to it.

def term_to_int(cell_contents):
  return int(cell_contents.strip().strip(' months'))

In [0]:
# Test out our function by calling it on our example
term_to_int(term)

60

In [0]:
Apply our solution to every cell in a column
df['term'] = df['term'].apply(term_to_int)

In [0]:
#Lets check out work
#Has the term column been cleaned up?
#Did the rule apply to ever cell in the column?
df.head()

loan_amnt  funded_amnt  funded_amnt_inv  term  int_rate  installment grade  \
0      20000        20000          20000.0    36     14.47       688.13     C   
1      16000        16000          16000.0    36     14.47       550.51     C   
2       8425         8425           8425.0    36     27.27       345.18     E   
3      10000        10000          10000.0    60     12.98       227.43     B   
4      30000        30000          30000.0    60     26.31       903.73     E   

  sub_grade                        emp_title emp_length home_ownership  \
0        C2                       Bus Driver    4 years            OWN   
1        C2                        Carpenter  10+ years           RENT   
2        E5  Senior Director Risk Management    3 years       MORTGAGE   
3        B5                          Printer    9 years           RENT   
4        E4                      Casino Host  10+ years            OWN   

   annual_inc verification_status    issue_d  loan_status pymnt_plan  \
0     52000.0     Source Verified 2018-12-01  Charged Off          n   
1    150000.0     Source Verified 2018-12-01      Current          n   
2    450000.0            Verified 2018-12-01   Fully Paid          n   
3     60000.0        Not Verified 2018-12-01      Current          n   
4     62000.0     Source Verified 2018-12-01  Charged Off          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    681xx         NE  30.65   
1               other                    Other    917xx         CA  17.94   
2         credit_card  Credit card refinancing    020xx         MA  12.37   
3         credit_card  Credit card refinancing    660xx         KS  14.90   
4  debt_consolidation       Debt consolidation    920xx         CA  37.24   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            1       1979-06-01               2                    15.0   
1            0       2001-11-01               1                     NaN   
2            0       1997-10-01               0                     NaN   
3            0       2007-05-01               0                     NaN   
4            0       2007-10-01               0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0      15048        73%         22   
1                     NaN        24        0      43721      59.2%         29   
2                     NaN        21        0      36812      65.7%         37   
3                   112.0         7        1      10677      54.2%         12   
4                    86.0        15        1      18890      39.8%         36   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w       0.00           0.00  8169.130000          8169.13   
1                   w    9726.53        9726.53  8788.870000          8788.87   
2                   w       0.00           0.00  8735.148975          8735.15   
3                   w    7787.21        7787.21  3805.020000          3805.02   
4                   w       0.00           0.00  2645.410000          2645.41   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          5731.54        2437.59                 0.0         0.0   
1          6273.47        2515.40                 0.0         0.0   
2          8425.00         310.15                 0.0         0.0   
3          2212.79        1592.23                 0.0         0.0   
4           754.24        1891.17                 0.0         0.0   

   collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d  \
0                      0.0     Feb-2020             0.00          NaN   
1                      0.0     May-2020             0.00     Jun-2020   
2                      0.0     Feb-2019          8754.30          NaN   
3                      0.0     May-2020    

#Create a function to apply the rule:
> Under the loan_status column, is the status is "Current" or "Fully Paid" the data needs to reflect that as the number 1. 

>Otherwise, the data in that same column needs to return with the number 0.

>After creating the function you will need to apply that rule to every cell within that column ('loan_status')
>You will need to use the .apply() function. We can reuse that function on those individual cells of a dataframe column via the .apply() function.

*Be aware of when and when not to use the quotes* **bold text**

In [0]:
def change_status(loan_status):
  if loan_status == 'Current' or loan_status == 'Fully Paid':
    return 1
  else:
    return 0

df['loan_status'] = df['loan_status'].apply(change_status)

In [0]:
df.head()

#Check your work.

loan_amnt  funded_amnt  funded_amnt_inv  term  int_rate  installment grade  \
0      20000        20000          20000.0    36     14.47       688.13     C   
1      16000        16000          16000.0    36     14.47       550.51     C   
2       8425         8425           8425.0    36     27.27       345.18     E   
3      10000        10000          10000.0    60     12.98       227.43     B   
4      30000        30000          30000.0    60     26.31       903.73     E   

  sub_grade                        emp_title emp_length home_ownership  \
0        C2                       Bus Driver    4 years            OWN   
1        C2                        Carpenter  10+ years           RENT   
2        E5  Senior Director Risk Management    3 years       MORTGAGE   
3        B5                          Printer    9 years           RENT   
4        E4                      Casino Host  10+ years            OWN   

   annual_inc verification_status    issue_d  loan_status pymnt_plan  \
0     52000.0     Source Verified 2018-12-01            0          n   
1    150000.0     Source Verified 2018-12-01            1          n   
2    450000.0            Verified 2018-12-01            1          n   
3     60000.0        Not Verified 2018-12-01            1          n   
4     62000.0     Source Verified 2018-12-01            0          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    681xx         NE  30.65   
1               other                    Other    917xx         CA  17.94   
2         credit_card  Credit card refinancing    020xx         MA  12.37   
3         credit_card  Credit card refinancing    660xx         KS  14.90   
4  debt_consolidation       Debt consolidation    920xx         CA  37.24   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            1       1979-06-01               2                    15.0   
1            0       2001-11-01               1                     NaN   
2            0       1997-10-01               0                     NaN   
3            0       2007-05-01               0                     NaN   
4            0       2007-10-01               0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0      15048        73%         22   
1                     NaN        24        0      43721      59.2%         29   
2                     NaN        21        0      36812      65.7%         37   
3                   112.0         7        1      10677      54.2%         12   
4                    86.0        15        1      18890      39.8%         36   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w       0.00           0.00  8169.130000          8169.13   
1                   w    9726.53        9726.53  8788.870000          8788.87   
2                   w       0.00           0.00  8735.148975          8735.15   
3                   w    7787.21        7787.21  3805.020000          3805.02   
4                   w       0.00           0.00  2645.410000          2645.41   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          5731.54        2437.59                 0.0         0.0   
1          6273.47        2515.40                 0.0         0.0   
2          8425.00         310.15                 0.0         0.0   
3          2212.79        1592.23                 0.0         0.0   
4           754.24        1891.17                 0.0         0.0   

   collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d  \
0                      0.0     Feb-2020             0.00          NaN   
1                      0.0     May-2020             0.00     Jun-2020   
2                      0.0     Feb-2019          8754.30          NaN   
3                      0.0     May-2020    

In [0]:
#Make last_pymnt_d_month and last_pymnt_d_year columns.

df['last_pymnt_d'] = pd.to_datetime(df['last_pymnt_d'], infer_datetime_format=True)

Now we can see that the 'last_pymnt_d' column has been changed to hold datetime objects.

Lets look at one of the cells specifically to see what a datetime object looks like:

In [0]:
df['last_pymnt_d'].head()

0   2020-02-01
1   2020-05-01
2   2019-02-01
3   2020-05-01
4   2019-04-01
Name: last_pymnt_d, dtype: datetime64[ns]

In [0]:
# Instead of a format like this example:
# Oct-2018





df['last_pymnt_d_month'] = df['last_pymnt_d'].dt.month

In [0]:
df['last_pymnt_d_year'] = df['last_pymnt_d'].dt.year

In [0]:
df['last_pymnt_d_year'][0]

# You can see how the month and year have been indicated by the strings that were
# contained in the column previously, and that the rest of the values have been 
# inferred.

In [0]:
# Let's check to see if the two columns were added
# AND
# are formatted as we needed. 
df.head()

loan_amnt  funded_amnt  funded_amnt_inv  term  int_rate  installment grade  \
0      20000        20000          20000.0    36     14.47       688.13     C   
1      16000        16000          16000.0    36     14.47       550.51     C   
2       8425         8425           8425.0    36     27.27       345.18     E   
3      10000        10000          10000.0    60     12.98       227.43     B   
4      30000        30000          30000.0    60     26.31       903.73     E   

  sub_grade                        emp_title emp_length home_ownership  \
0        C2                       Bus Driver    4 years            OWN   
1        C2                        Carpenter  10+ years           RENT   
2        E5  Senior Director Risk Management    3 years       MORTGAGE   
3        B5                          Printer    9 years           RENT   
4        E4                      Casino Host  10+ years            OWN   

   annual_inc verification_status    issue_d  loan_status pymnt_plan  \
0     52000.0     Source Verified 2018-12-01            0          n   
1    150000.0     Source Verified 2018-12-01            1          n   
2    450000.0            Verified 2018-12-01            1          n   
3     60000.0        Not Verified 2018-12-01            1          n   
4     62000.0     Source Verified 2018-12-01            0          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    681xx         NE  30.65   
1               other                    Other    917xx         CA  17.94   
2         credit_card  Credit card refinancing    020xx         MA  12.37   
3         credit_card  Credit card refinancing    660xx         KS  14.90   
4  debt_consolidation       Debt consolidation    920xx         CA  37.24   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            1       1979-06-01               2                    15.0   
1            0       2001-11-01               1                     NaN   
2            0       1997-10-01               0                     NaN   
3            0       2007-05-01               0                     NaN   
4            0       2007-10-01               0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0      15048        73%         22   
1                     NaN        24        0      43721      59.2%         29   
2                     NaN        21        0      36812      65.7%         37   
3                   112.0         7        1      10677      54.2%         12   
4                    86.0        15        1      18890      39.8%         36   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w       0.00           0.00  8169.130000          8169.13   
1                   w    9726.53        9726.53  8788.870000          8788.87   
2                   w       0.00           0.00  8735.148975          8735.15   
3                   w    7787.21        7787.21  3805.020000          3805.02   
4                   w       0.00           0.00  2645.410000          2645.41   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          5731.54        2437.59                 0.0         0.0   
1          6273.47        2515.40                 0.0         0.0   
2          8425.00         310.15                 0.0         0.0   
3          2212.79        1592.23                 0.0         0.0   
4           754.24        1891.17                 0.0         0.0   

   collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d  \
0                      0.0   2020-02-01             0.00          NaN   
1                      0.0   2020-05-01             0.00     Jun-2020   
2                      0.0   2019-02-01          8754.30          NaN   
3                      0.0   2020-05-01    

# Stretch Goals

You can do more with the LendingClub or Instacart datasets.

LendingClub options:
- There's one other column in the dataframe with percent signs. Remove them and convert to floats. You'll need to handle missing values.
- Modify the `emp_title` column to replace titles with 'Other' if the title is not in the top 20. 
- Take initiatve and work on your own ideas!

Instacart options:
- Read [Instacart Market Basket Analysis, Winner's Interview: 2nd place, Kazuki Onodera](http://blog.kaggle.com/2017/09/21/instacart-market-basket-analysis-winners-interview-2nd-place-kazuki-onodera/), especially the **Feature Engineering** section. (Can you choose one feature from his bulleted lists, and try to engineer it with pandas code?)
- Read and replicate parts of [Simple Exploration Notebook - Instacart](https://www.kaggle.com/sudalairajkumar/simple-exploration-notebook-instacart). (It's the Python Notebook with the most upvotes for this Kaggle competition.)
- Take initiative and work on your own ideas!

You can uncomment and run the cells below to re-download and extract the Instacart data

In [0]:
# !wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

In [0]:
# !tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

In [0]:
# %cd instacart_2017_05_01